registry/quilt_server/views.py

```diff
@@ -2307,6 +2316,174 @@ def _comment_dict(comment):
         contents=comment.contents
     )
 
+@app.route('/api/users/set_role', methods=['POST'])
+@api(enabled=ENABLE_USER_ENDPOINTS, require_admin=True, schema=USERNAME_ROLE_SCHEMA)
+@as_json
+def set_role():
+    """
+    Manages the role attached to a user.
+
+    The request should contain a JSON object with two keys:
+        username(string): username of user to edit
+        role(string): name of role to attach to user
+
+    A user can only have one role at a time.
+    To remove a role from a user, set role_name to the empty string.
+    """
+    data = request.get_json()
+    username = data['username']
+    role_name = data['role']
+    user = User.query.filter_by(name=username).one_or_none()
+    if user is None:
+        raise ApiException(requests.codes.bad_request,
+                           "No user exists by the provided name.")
+    if role_name == '':
+        # delete role from user
+        user.role_id = None
+    else:
+        role = Role.query.filter_by(name=role_name).one_or_none()
+        if role is None:
+            raise ApiException(requests.codes.bad_request,
+                               "No role exists by the provided name.")
+        user.role_id = role.id
+
+    db.session.add(user)
+    db.session.commit()
+    return {}
+
+def _role_dict(role):
+    role_dict = {
+        'id': role.id,
+        'name': role.name,
+        'arn': role.arn
+    }
+    return role_dict
+
+@app.route('/api/roles', methods=['POST'])
+@api(enabled=ENABLE_USER_ENDPOINTS, require_admin=True, schema=ROLE_DETAILS_SCHEMA)
+@as_json
+def add_role():
+    """
+    The body of the request should contain a JSON object.
+    There is one required parameter:
+        name(string): name of the role to operate on
+
+    There is an optional paramters:
+        arn(string): ARN of the IAM role associated with the Quilt role.
+
+    To create a role, you must provide an unused name and an arn.
+    """
+    data = request.get_json()
+    role_name = data['name']
+    arn = data.get('arn', None)
+    role = Role.query.filter_by(name=role_name).one_or_none()
+    if role is None:
+        if arn is None:
+            raise ApiException(
+                requests.codes.bad_request,
+                "Creating a role requires a role ARN"
+                )
+        if not VALID_NAME_RE.match(role_name):
```

Comment:
Minor, but it'd be better to lift this check outside the database transaction.

Reply:
Done